src/main/clojure/clojure/tools/gitlibs/impl.cljr

```diff
@@ -132,14 +203,96 @@
+              (do
+                (.Kill proc)
+                (throw (ex-info "Clone timed out (lock expired)" {:url url :git-path git-path})))
+              (do
+                (Thread/Sleep 1000)
+                (recur))))))
+      (catch Exception e
+        (try
+          (.Kill proc)
+          (catch Exception _))
+        (throw e)))))
+
+(defn- wait-for-clone-completion
+  "Waits for the clone to complete by checking if the git dir config file exists.
+   Updates the lock file timestamp every second while waiting.
+   Throws if the lock file timestamp expires."
+  [^String config-file-path ^String lockfile timeout-seconds]
+  (loop []
+    (when-not (System.IO.File/Exists config-file-path)
+      (write-ts lockfile)
+      (if (timestamp-expired? lockfile timeout-seconds)
+        (throw (ex-info "Cloner process dead (lock expired)" {}))
+        (do
+          (Thread/Sleep 1000)
+          (recur))))))
+
 (defn ensure-git-dir
-  "Ensure the bare git dir for the specified url, return the path to the git dir."
+  "Ensure the bare git dir for the specified url, return the path to the git dir.
+   
+   Uses lock-file-based coordination to prevent race conditions when multiple processes
+   attempt to clone the same repository simultaneously. The algorithm:
+   
+   1. Fast path: If git dir exists and no lock file, return immediately
+   2. Try to acquire lock by atomically creating lock file
+   3. If lock acquired: clone repo, update timestamp, release lock
+   4. If lock held by another process: wait for clone, update timestamp, check for expiry
+   
+   The lock file contains a timestamp that both the cloner and waiting processes update
+   every second. If the timestamp becomes >10 seconds old, it indicates a dead process
+   and an exception is thrown."
   [url]
   (let [git-dir-file (git-dir url)
-        config-file (cio/file-info git-dir-file "config")]                                        ;;; jio/file
-    (when-not (.Exists config-file)                                                               ;;; .exists
-      (git-clone-bare url git-dir-file))
-    (.FullName git-dir-file)))                                                                    ;;; .getCanonicalPath
+        config-file (cio/file-info git-dir-file "config")
+        config-file-path (.FullName config-file)
+        lockfile (str (.FullName git-dir-file) ".lock")
+        timeout-seconds 10]
+    ;; Fast path: if config exists and no lock file, we're done
+    (if (and (.Exists config-file) (not (System.IO.File/Exists lockfile)))
+      (.FullName git-dir-file)
+      ;; Otherwise, coordinate the clone
+      (if (acquire-lock lockfile)
+        ;; We acquired the lock - we are responsible for cloning
+        (try
+          (when-not (.Exists config-file)
+            (clone-with-timestamp-updates url git-dir-file lockfile timeout-seconds))
+          (.FullName git-dir-file)
+          (finally
+            (release-lock lockfile)))
```

Comment:
We need to add a catch block that ensures the git dir is deleted. In normal execution, we release the lock, but if thrown we need to delete the git dir first then the lock.